File: sdks/typescript/src/delivery_values.ts
/** Durable delivery outcome/phase values projected without owning the ledger. */
export const DeliveryOutcome = Object.freeze({
  Completed: "completed",
  Failed: "failed",
  Cancelled: "cancelled",
  Dropped: "dropped",
  OutcomeUnknown: "outcome_unknown",
} as const);

export type DeliveryOutcome = (typeof DeliveryOutcome)[keyof typeof DeliveryOutcome];

export function deliveryOutcomeAsStr(outcome: DeliveryOutcome): string {
  if (!Object.values(DeliveryOutcome).includes(outcome)) {
    throw new TypeError("invalid delivery outcome");
  }
  return outcome;
}

export const DeliveryPhase = Object.freeze({
  Persisted: "persisted",
  Claimed: "claimed",
  EffectStarted: "effect_started",
  MailboxAccepted: "mailbox_accepted",
  Drained: "drained",
  Deferred: "deferred",
  TurnSettled: "turn_settled",
} as const);

export type DeliveryPhase = (typeof DeliveryPhase)[keyof typeof DeliveryPhase];

export function deliveryPhaseAsStr(phase: DeliveryPhase): string {
  if (!Object.values(DeliveryPhase).includes(phase)) {
    throw new TypeError("invalid delivery phase");
  }
  return phase;
}

File: sdks/python/src/echo_agent_sdk/delivery_values.py
from __future__ import annotations

from enum import Enum


class DeliveryOutcome(str, Enum):
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"
    DROPPED = "dropped"
    OUTCOME_UNKNOWN = "outcome_unknown"

    def as_str(self) -> str:
        return self.value


class DeliveryPhase(str, Enum):
    PERSISTED = "persisted"
    CLAIMED = "claimed"
    EFFECT_STARTED = "effect_started"
    MAILBOX_ACCEPTED = "mailbox_accepted"
    DRAINED = "drained"
    DEFERRED = "deferred"
    TURN_SETTLED = "turn_settled"

    def as_str(self) -> str:
        return self.value
